mockgen/model/model_gotypes.go

```diff
@@ -0,0 +1,161 @@
+package model
+
+import (
+	"fmt"
+	"go/types"
+)
+
+// InterfaceFromGoTypesType returns a pointer to an interface for the
+// given reflection interface type.
+func InterfaceFromGoTypesType(it *types.Interface) (*Interface, error) {
+	intf := &Interface{}
+
+	for i := 0; i < it.NumMethods(); i++ {
+		mt := it.Method(i)
+		// Skip unexported methods.
+		if !mt.Exported() {
+			continue
+		}
+		m := &Method{
+			Name: mt.Name(),
+		}
+
+		var err error
+		m.In, m.Variadic, m.Out, err = funcArgsFromGoTypesType(mt.Type().(*types.Signature))
+		if err != nil {
+			return nil, err
```

Comment:
Given that we're descending into a tree, we may want to add context for sub-steps that can fail.
```suggestion
			return nil, fmt.Errorf("method %q: %w", mt.Name(), err)
```